### PR TITLE
Add support for BDD features / stories

### PR DESCRIPTION
**Pull request by leexuan0128**
Hi there,

This pull request adds the support for BDD (behavior-driven development) features/stories discussed in issue #668.  
We added keywords, punctuations, comments and numbers and etc for lexers to parse. It now should highlight those different parts correctly. Let us know if there are any bugs that need to be fixed or somewhere that needs to be changed.

Test sample

### CODE REVIEW

**Comment by kirito330824** [comment]
> Hi there,
> 
> This pull request adds the support for BDD (behavior-driven development) features/stories discussed in issue #668.
> We added keywords, punctuations, comments and numbers and etc for lexers to parse. It now should highlight those different parts correctly. Let us know if there are any bugs that need to be fixed or somewhere that needs to be changed.

Thank you, any reply will be appreciated.

**Comment by Anteru** [comment]
Hi, thanks for preparing this. A few notes:
* Please add a test file with a known-good output.
* Don't change the .gitignore
* Please remove the .ds_store files which are part of this commit

**Comment by kirito330824** [comment]
> Hi, thanks for preparing this. A few notes:
> 
> * Please add a test file with a known-good output.
> * Don't change the .gitignore
> * Please remove the .ds_store files which are part of this commit

Thanks for your reply. We will fix it base on your suggestion.

**Comment by leexuan0128** [comment]
> Hi, thanks for preparing this. A few notes:
> 
> * Please add a test file with a known-good output.
> * Don't change the .gitignore
> * Please remove the .ds_store files which are part of this commit

Hi, thanks for your reminder, already add test cases. let me know if there are any changes that need to make :)

File: pygments/lexers/bdd.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    :copyright: Copyright 2006-2021 by the Pygments team, see AUTHORS.
+    :license: BSD, see LICENSE for details.
+"""
+
+from pygments.lexer import RegexLexer, include, bygroups
+from pygments.token import Comment, Keyword, Name, String, Number, Generic, Text
+
+__all__ = ['BddLexer']
+
+class BddLexer(RegexLexer):
```
Comment (Anteru) [inline pygments/lexers/bdd.py:16]: Needs a `.. versionadded: 2.10` comment (and in general a comment explaining what this lexer is about.)

File: pygments/lexers/bdd.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+            (r"\\\|", String),
+            (r"\s*\|", Keyword),
+            include('string'),
+        ],
+        'double_string': [
+            (r'"', Name.Function, "#pop"),
+            include('string'),
+        ],
+        'root': [
+            (r'\n', Name.Function),
```
Comment (Anteru) [inline pygments/lexers/bdd.py:108]: Are you sure about this? This looks like it should be whitespace unless I'm missing something here.

File: pygments/lexers/bdd.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+             'step_content_root'),
+            (feature_keywords, bygroups(Keyword, Keyword, Name.Function),
+             'narrative'),
+            (feature_element_keywords,
+             bygroups(Name.Function, Keyword, Keyword, Name.Function),
+             'feature_elements'),
+            (examples_keywords,
+             bygroups(Name.Function, Keyword, Keyword, Name.Function),
+             'examples_table'),
+            (r'(\s|.)', Name.Function),
```
Comment (Anteru) [inline pygments/lexers/bdd.py:124]: This should be whitespace as well, and it also has a catch-all -- is that intended?

File: pygments/lexers/bdd.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+            (r'(:)', Keyword.Type),
+            (r'((?<=\<)[^\\>]+(?=\>))', Name.Variable),
+            (r'"([^\"]*)"', Name.Exception),
+        ],
+        'numbers': [
+            (r'(\d+\.?\d*|\d*\.\d+)([eE][+-]?[0-9]+)?', String),
+        ],
+        'string': [
+            include('table_vars'),
+            (r'(\s|.)', String),
```
Comment (Anteru) [inline pygments/lexers/bdd.py:76]: Same as below.

File: pygments/lexers/bdd.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        ],
+        'scenario_sections_on_stack': [
+            (feature_element_keywords,
+             bygroups(Name.Function, Keyword, Keyword, Name.Function),
+             "feature_elements_on_stack"),
+        ],
+        'narrative': [ # narrative text
+            include('scenario_sections_on_stack'),
+            include('comments'),
+            (r"(\s|.)", Name.Function),
```
Comment (Anteru) [inline pygments/lexers/bdd.py:62]: Same as below

File: pygments/lexers/bdd.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    step_keywords = '^(\\s*)(Given|When|Then|Add|And|Feature|Scenario Outline|Scenario|Background|Examples|But|\\* )'
+
+    tokens = {
+        'comments': [
+            (r'^\s*#.*$', Comment),
+        ],
+        'feature_elements': [
+            (step_keywords, Keyword, "step_content_stack"),
+            include('comments'),
+            (r"(\s|.)", Name.Function),
```
Comment (Anteru) [inline pygments/lexers/bdd.py:35]: Same as below

File: pygments/lexers/bdd.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+        ],
+        'feature_elements_on_stack': [
+            (step_keywords, Keyword, "#pop:2"),
+            include('comments'),
+            (r"(\s|.)", Name.Function),
```
Comment (Anteru) [inline pygments/lexers/bdd.py:40]: Same as below

File: pygments/lexers/bdd.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+        ],
+        'step_content_stack': [
+            (r"$", Keyword, "#pop:2"),
+            include('step_content'),
+        ],
+        'step_content': [
+            include('table_vars'),
+            include('numbers'),
+            include('comments'),
+            (r'(\s|.)', Name.Function),
```
Comment (Anteru) [inline pygments/lexers/bdd.py:94]: Same as below

**Review by Anteru — CHANGES_REQUESTED** [review]

File: pygments/lexers/bdd.py (hunk 1 above)
Comment (leexuan0128) [reply pygments/lexers/bdd.py:17]: Done. Thanks for your reminder 😃

File: pygments/lexers/bdd.py (hunk 2 above)
Comment (leexuan0128) [reply pygments/lexers/bdd.py]: Oh I made a stupid mistake here lol

File: pygments/lexers/bdd.py (hunk 1 above)
Comment (leexuan0128) [reply pygments/lexers/bdd.py:17]: Done.

Comment (leexuan0128) [reply pygments/lexers/bdd.py:17]: Done.

**Review by leexuan0128 — COMMENTED** [review]
add `versionadded: 2.10` comment and the explanation

File: pygments/lexers/bdd.py (hunk 3 above)
Comment (leexuan0128) [reply pygments/lexers/bdd.py]: yes, it should be whitespace here. thanks for your correction.

**Comment by Anteru** [comment]
Changes look mostly good, but you do highlight whitespace as `Text`. Please assign whitespace to things like ` ` and `\n` as some style might highlight whitespace differently from text. Check the golden test output.

File: pygments/lexers/bdd.py (hunk 3 above)
Comment (leexuan0128) [reply pygments/lexers/bdd.py]: It should show the '\n'  and '  ' as the whitespace now 😃

File: tests/examplefiles/bdd/example.feature.output [hunk 9]
```diff
@@ -0,0 +1,719 @@
...
+' '           Text.Whitespace
+'G'           Text
+'989'         Literal.Number
+'\n'          Text.Whitespace
+
+'\n'          Text.Whitespace
+
+'Feature'     Keyword
+':'           Punctuation
+' '           Text
```
Comment (amitkummer) [inline tests/examplefiles/bdd/example.feature.output:54]: Should be `Whitespace`?

File: tests/examplefiles/bdd/example.feature.output [hunk 10]
```diff
@@ -0,0 +1,719 @@
...
+'Background'  Keyword
+':'           Punctuation
+''            Keyword
+'\n'          Text.Whitespace
+
+'    '        Text.Whitespace
+'Given'       Keyword
+' '           Text
+'a'           Text
+' '           Text
```
Comment (amitkummer) [inline tests/examplefiles/bdd/example.feature.output:259]: Should be `Whitespace`.

File: pygments/lexers/bdd.py [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+    step_keywords = '^(\\s*)(Given|When|Then|Add|And|Feature|Scenario Outline|Scenario|Background|Examples|But)'
+
+    tokens = {
+        'comments': [
+            (r'^\s*#.*$', Comment),
+        ],
+        'feature_elements': [
+            (step_keywords, Keyword, "step_content_stack"),
+            include('comments'),
+            (r"(\s)", Whitespace),
```
Comment (amitkummer) [inline pygments/lexers/bdd.py:43]: Any reason for using a group match here? Also maybe consider capturing multiple spaces for efficiency?
```suggestion
            (r"\s+", Whitespace),
```

File: pygments/lexers/bdd.py [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+        ],
+        'narrative': [
+            include('scenario_sections_on_stack'),
+            include('comments'),
+            (r"(\s)", Whitespace),
+        ],
+        'table_vars': [
+            (r'(<|>)', Punctuation),
+            (r'(\|)', Punctuation),
+            (r'(:)', Punctuation),
```
Comment (amitkummer) [inline pygments/lexers/bdd.py:75]: Merge this into a single rule maybe?

File: pygments/lexers/bdd.py [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+        ],
+        'step_content_stack': [
+            (r"$", Keyword, "#pop:2"),
+            include('step_content'),
+        ],
+        'step_content': [
+            include('table_vars'),
+            include('numbers'),
+            include('comments'),
+            (r'(\s|.)', Text),
```
Comment (amitkummer) [inline pygments/lexers/bdd.py:102]: This matches spaces as `Text`.

File: pygments/lexers/bdd.py [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+        ],
+        'table_content': [
+            (r"\s+\|\s*$", Punctuation, "#pop"),
+            include('comments'),
+            (r"\\\|", Punctuation),
+            (r"\s*\|", Punctuation),
```
Comment (amitkummer) [inline pygments/lexers/bdd.py:108]: Spaces for `Punctuation`.

File: pygments/lexers/bdd.py [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+            include('string'),
+        ],
+        'root': [
+            (r'\n', Whitespace),
```
Comment (amitkummer) [inline pygments/lexers/bdd.py:112]: Merge this rule with the `(r'\s', Whitespace),` one at the bottom?

File: pygments/lexers/bdd.py [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+            include('comments'),
+            (r'"""', Keyword, "py_string"),
+            include('table_vars'),
+            include('numbers'),
+            (r'(\s*)(@[^@\s]+)', bygroups(Text, Name.Label)),
```
Comment (amitkummer) [inline pygments/lexers/bdd.py:117]: The first group which is spaces is matched as `Text`.

**Review by amitkummer — CHANGES_REQUESTED** [review]
Thanks a lot for your work! Some of the things I commented about appear on multiple places, so please, if you think my comment/suggestion is justified, then change it in all other places as well.

Also, check your test output, it has issues (I see spaces as `Text` a lot).

File: pygments/lexers/bdd.py (hunk 11 above)
Comment (leexuan0128) [reply pygments/lexers/bdd.py]: Thank you very much for your patient guidance. I overcomplicated the problem. I delete the unnecessary parts. Now it is more concise. 😃

File: pygments/lexers/bdd.py [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+        'numbers': [
+            (r'(\d+\.?\d*|\d*\.\d+)([eE][+-]?[0-9]+)?', Number),
+        ],
+        'root': [
+            (r'\n|\s', Whitespace),
+            (step_keywords, Keyword),
+            include('comments'),
+            include('miscellaneous'),
+            include('numbers'),            
+            (r'.', Text),
```
Comment (Anteru) [inline pygments/lexers/bdd.py:51]: This rule produces tons of tokens each containing a single character. I think you want `\S+` here, which should also dramatically cut down the size of your test output.

**Review by Anteru — CHANGES_REQUESTED** [review]

File: pygments/lexers/bdd.py (hunk 17 above)
Comment (kirito330824) [reply pygments/lexers/bdd.py]: Thank you so much for your advice. We have changed the bdd.py according to your help. Now the token size was greatly reduced.

File: pygments/lexers/bdd.py [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+            (r'(<|>|\[|\]|=|\||:|\(|\)|\{|\}|,|\.|;|-|_|\$)', Punctuation),
+            (r'((?<=\<)[^\\>]+(?=\>))', Name.Variable),
+            (r'"([^\"]*)"', String),
+            (r'^@\S+', Name.Label),
+        ],
+        'numbers': [
+            (r'(\d+\.?\d*|\d*\.\d+)([eE][+-]?[0-9]+)?', Number),
+        ],
+        'root': [
+            (r'\n|\s', Whitespace),
```
Comment (Anteru) [inline pygments/lexers/bdd.py:46]: That should be `\s+` so you don't produce a new token for each space (sorry I missed this previously.)

**Review by Anteru — CHANGES_REQUESTED** [review]

File: pygments/lexers/bdd.py (hunk 18 above)
Comment (kirito330824) [reply pygments/lexers/bdd.py]: Thanks for your correction. Now, bdd.py won't generate the new token for each space.

**Comment by Anteru** [comment]
Merged, sorry it took so long.

**Pull request merged**